### PR TITLE
refactor: create a user service

### PR DESCRIPTION
**Pull request by AlexKMarshall**
This is a refactor PR that attempts to make a more testable set of patterns. The signup route action is refactored to use a user service and a user repository, allowing the user repository to be mocked out, but let everything else (for instance password hashing) running.
In further PRs will update the login functions to also use these patterns

## Greptile Summary

This PR refactors the signup functionality to implement the repository pattern with dependency injection, significantly improving testability and code organization. The changes introduce a layered architecture where business logic is separated from data access concerns.

### Key Changes:
• **Created `UserService`** - Handles business logic including password hashing and user creation workflow
• **Implemented `IUserRepository` interface** - Enables dependency injection and mocking for tests
• **Added `UserRepository` class** - Manages database operations using Drizzle ORM
• **Extracted password utilities** - Moved `hashPassword` function to dedicated utility module
• **Created Database type** - Centralized type definition for consistent database typing
• **Updated tests and stories** - Refactored to use new service/repository pattern

The refactor follows SOLID principles and clean architecture patterns, moving from direct database calls to a structured service layer. This enables mocking of database operations in tests while keeping business logic (like password hashing) intact and testable. The changes lay the foundation for similar patterns to be applied to login functionality in future PRs.

## Important Files Changed

<details>
<summary>Click to expand file changes</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| `app/features/user/user-service.ts` | 4/5 | New service and repository classes implementing dependency injection pattern with proper password hashing |
| `app/auth/signup.tsx` | 4/5 | Refactored to use UserService instead of direct database calls, enabling better testability |
| `app/auth/_actions/signup-action.ts` | 4/5 | Updated to accept UserService dependency, maintaining business logic while improving test isolation |
| `app/auth/_actions/signup-action.test.ts` | 4/5 | Comprehensive test refactor using mock repository pattern for isolated unit testing |
| `app/utils/password.ts` | 4/5 | Extracted password hashing utility from data layer into reusable module with strong cryptographic parameters |
| `database/index.ts` | 5/5 | Simple but critical Database type definition enabling dependency injection throughout the application |
| `app/auth/signup.stories.tsx` | 5/5 | Updated Storybook story to use new UserService pattern for consistent error testing |

</details>

## Confidence score: 4/5

- This PR introduces significant architectural improvements with proper dependency injection and testing patterns
- Score reflects excellent code organization and testability improvements, with minor concerns around error handling completeness
- Pay close attention to `app/features/user/user-service.ts` for missing duplicate email validation and error handling

## Manual Testing Recommendations

- [ ] Test complete signup flow with valid email/password combinations
- [ ] Test signup with duplicate email addresses to verify error handling
- [ ] Test signup with invalid inputs (malformed email, weak passwords)
- [ ] Verify password hashing is working correctly in database
- [ ] Test error scenarios including database connection failures
- [ ] Confirm redirect to homepage after successful signup
- [ ] Test form validation on blur and input events
- [ ] Verify DatabaseError story displays correctly in Storybook

## Automated Testing

- [ ] Add unit tests for UserService.signup method with various scenarios
- [ ] Add unit tests for UserRepository.create method with mocked database
- [ ] Add integration tests for complete signup flow with real database
- [ ] Add tests for duplicate email constraint handling
- [ ] Add password hashing verification and security tests
- [ ] Add contract tests to ensure IUserRepository interface compliance
- [ ] Add error handling tests for database failures
- [ ] Add performance tests for password hashing with high iteration count

## Architecture and Design Review

**Excellent architectural improvements:**
• **Repository Pattern Implementation** - Clean separation between service layer (business logic) and repository layer (data access)
• **Dependency Injection** - Proper use of interfaces enabling easy mocking and testing
• **SOLID Principles** - Follows dependency inversion principle with interface abstractions
• **Type Safety** - Comprehensive TypeScript typing throughout the service layer
• **Testability** - Enables isolated unit testing while preserving business logic integrity

**Areas for improvement:**
• **Error Handling** - Missing specific validation for duplicate emails and database constraint violations
• **Input Validation** - Service layer could benefit from additional input validation beyond schema validation
• **Incomplete Migration** - `verifyPassword` function still commented out in password utilities
• **Error Specificity** - Generic error messages don't provide user-friendly feedback for different failure scenarios

## Sequence Diagram
```mermaid
sequenceDiagram
    participant User
    participant SignupForm as "Signup Form"
    participant SignupAction as "Signup Action"
    participant UserService as "User Service"
    participant UserRepository as "User Repository"
    participant PasswordUtils as "Password Utils"
    participant Database as "Database"

    User->>SignupForm: "Fill email & password"
    User->>SignupForm: "Submit form"
    SignupForm->>SignupAction: "POST formData"
    SignupAction->>SignupAction: "Parse & validate with SignupSchema"
    
    alt Valid form data
        SignupAction->>UserService: "signup({ email, password })"
        UserService->>PasswordUtils: "hashPassword(password)"
        PasswordUtils-->>UserService: "passwordHash"
        UserService->>UserRepository: "create({ user: { email }, password: { hash } })"
        UserRepository->>Database: "INSERT into users table"
        Database-->>UserRepository: "Created user record"
        UserRepository-->>UserService: "User object"
        UserService-->>SignupAction: "User object"
        SignupAction-->>SignupForm: "Redirect to /"
        SignupForm-->>User: "Navigate to homepage"
    else Invalid form data
        SignupAction-->>SignupForm: "Validation errors"
        SignupForm-->>User: "Display field errors"
    else Database error
        UserRepository->>Database: "INSERT fails"
        Database-->>UserRepository: "Error"
        UserRepository-->>UserService: "Throw error"
        UserService-->>SignupAction: "Throw error"
        SignupAction-->>SignupForm: "Form error message"
        SignupForm-->>User: "Display error message"
    end
```

### CODE REVIEW

File: app/auth/_actions/signup-action.test.ts [hunk 1]
```diff
@@ -1,13 +1,14 @@
 import { faker } from '@faker-js/faker'
 import { assert, expect, test, vi } from 'vitest'
 import { makeSignupAction } from './signup-action'
+import { UserService, type IUserRepository } from '~/features/user/user-service'
 
-const mockSaveUser = vi.fn()
-mockSaveUser.mockName('saveUser')
+const mockCreateUser = vi.fn()
+mockCreateUser.mockName('createUser')
```
[the remark this answers is not in the document]
Comment (AlexKMarshall) [reply app/auth/_actions/signup-action.test.ts:7]: This is implemented in the global vitest setup. Mocks are cleared for every test
